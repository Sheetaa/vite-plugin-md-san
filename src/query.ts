export interface Query {
    [index: string]: string
}

export interface RequestRes {
    filename: string;
    filepath: string;
    query?: Query;
}

export interface CodeLangRes {
    lang: string;
    export?: 'preview' | string;
    caption?: string;
}

export function parseRequest(id: string): RequestRes {
    const [filepath, rawQuery] = id.split('?', 2);
    const filename = filepath.split('/').pop() || '';
    if (!rawQuery) {
        return {
            filename,
            filepath
        };
    }
    const request = new URLSearchParams(rawQuery);
    const query = {} as Query;
    request.forEach((value, name) => {
        query[name] = value;
    });
    return {
        filename,
        filepath,
        query
    };
}

export function querystring(query: Query) {
    const result = [];
    for (let [key, value] of Object.entries(query)) {
        result.push(`${key}=${value ? value : ''}`);
    }
    return result.join('&');
}

export function parseCodeLang(id: string): CodeLangRes {
    const arr = id.split(' ');
    return arr.reduce((acc, cur, index) => {
        const pairs = cur.split('=');
        if (index === 0) {
            acc.lang = pairs[0];
        }
        else {
            let key: keyof CodeLangRes;
            key = pairs[0] as keyof CodeLangRes;
            if (pairs.length > 1) {
                acc[key] = pairs[1];
            }
        }
        return acc;
    }, {lang: ''} as CodeLangRes);
}
