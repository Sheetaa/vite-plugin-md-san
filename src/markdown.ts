/**
 * @file markdown compile
 * @author Sheeta(wuhayao@gmail.com)
 */

import fs from 'fs';
import path from 'path';
import {marked} from 'marked';
import {parseCodeLang} from './query';

type ExportType = 'html' | 'component';

interface ComponentSnippets {
    // import xxx from 'yyy';
    import: string;
    // 'xxx': xxx
    component: string;
}

interface TemplateData {
    id: string | number;
    // escaped code
    code: string;
    componentRequest: string;
    caption?: string;
}

interface CompileOptions {
    filepath: string;
    exportType?: ExportType;
    template?: string;
}

const defaultTemplate = path.join(__dirname, './theme/default.template');
let templateContent = fs.readFileSync(defaultTemplate, {encoding: 'utf8'});
let file: string;
let exportType: ExportType;
let index: number;
const components: ComponentSnippets[] = [];
const previewBlocks: Map<string, string> = new Map();

function init(options: CompileOptions) {
    file = options.filepath;
    exportType = options.exportType || 'html';
    templateContent = options.template || templateContent;
    index = 1;
    components.splice(0, components.length);
    previewBlocks.clear();
}

const renderer = {
    code(code: string, infostring: string) {
        const codeEsc = code
            .replace(/</g, '&lt;')
            .replace(/`/g, '&#96;');

        const codeLang = parseCodeLang(infostring);
        if (
            exportType === 'component'
            && codeLang.lang === 'san'
            && codeLang.export === 'preview'
        ) {
            const entryTag = `preview-block-${index}`;
            const entryVar = `PreviewBlock${index}`;
            const mapKeyEntry = `${entryVar}.vpms`;
            const mapKeyComponent = `Component${index}.vpms`;
            // /src/markdown/html.md.PreviewBlock1.vpms
            const entryRequest = `${file}.${mapKeyEntry}`;
            // /src/markdown/html.md.Component1.vpms
            const componentRequest = `${file}.${mapKeyComponent}`;
            components.push({
                import: `import ${entryVar} from '${entryRequest}'`,
                component: `'${entryTag}': ${entryVar}`
            });
            previewBlocks.set(mapKeyEntry, getTemplate({
                id: index,
                code: codeEsc,
                componentRequest,
                caption: codeLang.caption
            }));
            previewBlocks.set(mapKeyComponent, code);
            index++;

            return `<${entryTag}></${entryTag}>`;
        }

        return `<pre><code class="language-san">${codeEsc}</code></pre>`;
    }
};

marked.use({renderer});

export function compile(raw: string, compileOptions: CompileOptions) {
    init(compileOptions);

    const html = marked.parse(raw);
    if (compileOptions.exportType === 'html') {
        return {
            html
        };
    }

    const entryComponent = `import {Component} from 'san';
${components.map(item => item.import).join(';\n')};
export default class ComponentDoc extends Component {
    static template = \`<section class=\"markdown\">${html}</section>\`;
    static components = {
        ${components.map(item => item.component).join(',\n\t\t')}
    };
};
`;

    return {
        entryComponent,
        previewBlocks
    }
}

function getTemplate(data: TemplateData) {
    let template = templateContent;
    let key: keyof TemplateData;
    for (key in data) {
        const value = ('' + data[key]) ?? '';
        // <%=id=%>
        template = template.split(new RegExp(`<%=\\s*${key}\\s*=%>`, 'g')).join(value);
    }
    return template;
}
