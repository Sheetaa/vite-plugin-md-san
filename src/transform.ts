/**
 * @file transform function
 * @author Sheeta(wuhayao@gmail.com)
 */

import {parseRequest} from './query';
import {compile} from './markdown';
import {TransformOption} from './index';

export function createTransform(options: TransformOption) {

    return function (id: string, raw: string) {
        const {filepath, query} = parseRequest(id);
        const realOptions = Object.assign({
            export: 'component'
        }, options, query);

        const alias = options.config.resolve.alias || [];

        switch (realOptions.export) {
            case 'html':
                const {html} = compile(raw, {
                    filepath,
                    alias,
                    exportType: 'html'
                });
                return {
                    transformed: `export default \`${html}\``
                };
            case 'component':
                let {
                    entryComponent: transformed,
                    previewBlocks: attachment
                } = compile(raw, {
                    filepath,
                    exportType: 'component',
                    alias,
                    template: realOptions.template,
                    query
                });
                return {
                    transformed: transformed || '',
                    attachment
                };
            case 'raw':
            default:
                return {
                    transformed: `export default \`${raw}\``
                };
        }
    };
};
