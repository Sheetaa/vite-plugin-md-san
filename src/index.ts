/**
 * @file vite plugin markdown san
 * @author Sheeta(wuhayao@gmail.com)
 */

import path from 'path';
import {Plugin, ResolvedConfig, ModuleNode} from 'vite';
import {createFilter} from '@rollup/pluginutils';
import {createTransform} from './transform';
import {parseRequest} from './query';

const debug = require('debug')('vite-plugin-md-san:index');

export interface PluginOptions {
    export?: 'html' | 'component' | 'raw' | string;
    // todo
    template?: string;
    // todo
    highlight?: Record<string, unknown>;
}

let cachedPreviewBlocks: Map<string, Map<string, string>> = new Map();

export default function VitePluginMarkdownSan(options: PluginOptions): Plugin {

    const filter = createFilter(/\.md\??/);
    const previewFilter = createFilter(/\.vpms$/);
    const transform = createTransform(options);

    let config: ResolvedConfig;

    return {
        name: 'vite-plugin-md-san',
        enforce: 'pre',
        configResolved(resolvedConfig) {
            config = resolvedConfig
        },
        // first: resolve
        resolveId(source) {
            if (previewFilter(source)) {
                const id = source.startsWith(config.root)
                    ? source : path.join(config.root, source);
                return id;
            }
        },
        // second: transform
        transform(raw, id) {
            // filter file ending with .md
            if (!filter(id) || previewFilter(id)) {
                return;
            }

            const {filename} = parseRequest(id);

            const {
                transformed,
                attachment
            } = transform(id, raw);
            if (attachment && attachment.size) {
                cachedPreviewBlocks.set(filename, attachment);
            }
            return transformed;
        },
        // third: load
        load(id) {
            if (previewFilter(id)) {
                const filename = id.split('/').pop();
                const matched = filename?.match(/([\w-]+\.md)\.([\w-]+\.vpms)/);
                if (matched?.length === 3) {
                    const mdFilename = matched[1];
                    const previewBlockName = matched[2];
                    return cachedPreviewBlocks.get(mdFilename)?.get(previewBlockName);
                }
            }
        },
        async handleHotUpdate(ctx) {
            const {
                file,
                read
            } = ctx;

            if (!filter(file)) {
                return;
            }

            debug('handleHotUpdate:file', file);

            const {filename} = parseRequest(file);
            const defaultRead = read;
            ctx.read = async function() {
                const {
                    transformed,
                    attachment
                } = transform(file, await defaultRead());
                if (attachment && attachment.size) {
                    cachedPreviewBlocks.set(filename, attachment);
                }
                return transformed;
            };
        },
        buildEnd() {
            // todo
        },
        closeBundle() {
            // todo
        }
    };
}
