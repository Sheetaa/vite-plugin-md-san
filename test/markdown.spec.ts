/**
 * @file transform
 * @author Sheeta(wuhayao@gmail.com)
 */

import fs from 'fs';
import path from 'path';
import {compile} from '../src/markdown';

describe('transform', () => {

    test('html', () => {
        const md = `# title
123
456

\`\`\`san
import {Component} from 'san';

export default SanDemo extends Component {
    static template = \`<div></div>\`;
    static components = {};
    initData () {
        return {
            foo: 'bar'
        };
    }
}
\`\`\`
`;
        const {html} = compile(md, {
            filepath: '/home/work/project/foo.md',
            exportType: 'html'
        });
        expect(html).toMatchSnapshot();
    });

    test('component', () => {
        const relativePath = '../demo/src/markdown/san-include-preview.md';
        let filepath = path.resolve(__dirname, relativePath);
        const md = fs.readFileSync(filepath, {encoding: 'utf-8'});
        filepath = path.join('/home/work/project', relativePath);

        const {
            entryComponent,
            previewBlocks
        } = compile(md, {
            filepath,
            exportType: 'component'
        });
        expect(entryComponent).toMatchSnapshot();

        const keys = Array.from(previewBlocks || []).map(item => item[0]);
        expect(keys).toStrictEqual([
            'PreviewBlock1.vpms',
            'Component1.vpms',
            'PreviewBlock2.vpms',
            'Component2.vpms',
            'PreviewBlock3.vpms',
            'Component3.vpms'
        ]);
        expect(previewBlocks?.get('PreviewBlock1.vpms')).toMatchSnapshot();
        expect(previewBlocks?.get('Component1.vpms')).toMatchSnapshot();
    });

    test('component: custom preview template', () => {
        const relativePath = '../demo/src/markdown/san-include-preview.md';
        let filepath = path.resolve(__dirname, relativePath);
        const md = fs.readFileSync(filepath, {encoding: 'utf-8'});
        filepath = path.join('/home/work/project', relativePath);

        const customTemplatePath = path.resolve(__dirname, 'fixtures/example.template');
        const customTemplate = fs.readFileSync(customTemplatePath, {encoding: 'utf-8'});

        const {previewBlocks} = compile(md, {
            filepath,
            exportType: 'component',
            template: customTemplate
        });
        expect(previewBlocks?.get('PreviewBlock1.vpms')).toMatchSnapshot();
        expect(previewBlocks?.get('PreviewBlock2.vpms')).toMatchSnapshot();
    });

    test('component: custom preview template function', () => {
        const relativePath = '../demo/src/markdown/san-include-preview.md';
        let filepath = path.resolve(__dirname, relativePath);
        const md = fs.readFileSync(filepath, {encoding: 'utf-8'});
        filepath = path.join('/home/work/project', relativePath);

        const customTemplatePath = path.resolve(__dirname, 'fixtures/example.template');
        const customTemplate = fs.readFileSync(customTemplatePath, {encoding: 'utf-8'});

        const {previewBlocks} = compile(md, {
            filepath,
            exportType: 'component',
            template: function (data: any) {
                expect(data.filepath).toBe(filepath);
                return customTemplate;
            }
        });
        expect(previewBlocks?.get('PreviewBlock1.vpms')).toMatchSnapshot();
        expect(previewBlocks?.get('PreviewBlock2.vpms')).toMatchSnapshot();
    });

    test('component: custom preview template with more source', () => {
        const relativePath = '../demo/src/markdown/san-include-preview.md';
        let filepath = path.resolve(__dirname, relativePath);
        const md = fs.readFileSync(filepath, {encoding: 'utf-8'});
        filepath = path.join('/home/work/project', relativePath);

        const customTemplatePath = path.resolve(__dirname, 'fixtures/multi-file.template');
        const customTemplate = fs.readFileSync(customTemplatePath, {encoding: 'utf-8'});

        const {previewBlocks} = compile(md, {
            filepath,
            exportType: 'component',
            alias: [],
            template: customTemplate
        });
        expect(previewBlocks?.get('PreviewBlock1.vpms')).toMatchSnapshot();
        expect(previewBlocks?.get('PreviewBlock2.vpms')).toMatchSnapshot();
        expect(previewBlocks?.get('PreviewBlock3.vpms')).toMatchSnapshot();
    });
});
