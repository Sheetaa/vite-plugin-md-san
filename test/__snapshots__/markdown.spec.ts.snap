// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform component 1`] = `
"import {Component} from 'san';
import PreviewBlock1 from '/home/work/demo/src/markdown/san-include-preview.md.PreviewBlock1.vpms';
import PreviewBlock2 from '/home/work/demo/src/markdown/san-include-preview.md.PreviewBlock2.vpms';
import PreviewBlock3 from '/home/work/demo/src/markdown/san-include-preview.md.PreviewBlock3.vpms';
export default class ComponentDoc extends Component {
    static template = \`<section class=\\"markdown\\"><h1 id=\\"export-md-as-san-component\\">export md as san component</h1>
<preview-block-1></preview-block-1><preview-block-2></preview-block-2><preview-block-3></preview-block-3></section>\`;
    static components = {
        'preview-block-1': PreviewBlock1,
		'preview-block-2': PreviewBlock2,
		'preview-block-3': PreviewBlock3
    };
};
"
`;

exports[`transform component 2`] = `
"import {Component} from 'san';
import CodePreview from '/home/work/demo/src/markdown/san-include-preview.md.Component1.vpms';

export default class PreviewBlock extends Component {

    static trimWhitespace = 'all';

    static template = \`
        <section class=\\"code-box\\" id=\\"1\\">
            <section class=\\"code-box-demo\\">
                <h4>code preview</h4>
                <code-preview></code-preview>
            </section>
            <section class=\\"code-box-meta\\">
                <h4>info text</h4>
                {{ caption | raw}}
            </section>
            <section>
                <h4>code</h4>
                <div class=\\"highlight\\">
                    <pre>
                        <code class=\\"language-san\\">
                            {{ code | raw }}
                        </code>
                    </pre>
                </div>
            </section>
        </section>
    \`;

    static components = {
        'code-preview': CodePreview
    };

    initData() {
        return {
            code: \`import {Component} from 'san';

export default class ComponentDemo extends Component {
    static template = &#96;
        &lt;div>
            &lt;button type=\\"button\\">First Button&lt;/button>
        &lt;/div>
    &#96;;
}\`,
            caption: \`first\`
        }
    }
}
"
`;

exports[`transform component 3`] = `
"import {Component} from 'san';

export default class ComponentDemo extends Component {
    static template = \`
        <div>
            <button type=\\"button\\">First Button</button>
        </div>
    \`;
}"
`;

exports[`transform component: custom preview template 1`] = `
"import {Component} from 'san';
import CodePreview from '/home/work/demo/src/markdown/san-include-preview.md.Component1.vpms';

export default class PreviewBlock extends Component {

    static trimWhitespace = 'all';

    static template = \`
        <section class=\\"code-box\\" id=\\"1\\">
            <section class=\\"code-box-demo\\">
                <h4>code preview</h4>
                <code-preview></code-preview>
            </section>
        </section>
    \`;

    static components = {
        'code-preview': CodePreview
    };
}
"
`;

exports[`transform component: custom preview template 2`] = `
"import {Component} from 'san';
import CodePreview from '/home/work/demo/src/markdown/san-include-preview.md.Component2.vpms';

export default class PreviewBlock extends Component {

    static trimWhitespace = 'all';

    static template = \`
        <section class=\\"code-box\\" id=\\"2\\">
            <section class=\\"code-box-demo\\">
                <h4>code preview</h4>
                <code-preview></code-preview>
            </section>
        </section>
    \`;

    static components = {
        'code-preview': CodePreview
    };
}
"
`;

exports[`transform component: custom preview template with more source 1`] = `
"import {Component} from 'san';
import CodePreview from '/home/work/demo/src/markdown/san-include-preview.md.Component1.vpms';

export default class PreviewBlock extends Component {

    static trimWhitespace = 'all';

    static template = \`
        <section class=\\"code-box\\" id=\\"1\\">
            <section class=\\"code-box-demo\\">
                <h4>code preview</h4>
                <code-preview></code-preview>
            </section>
            <section class=\\"code-preview-wrapper\\">
                <section
                    class=\\"code-preview\\"
                    s-for=\\"source in sourceList\\"
                >
                    <h3>{{source.filename}}</h3>
                    <code s-html=\\"source.code\\"></code>
                </section>
            </section>
        </section>
    \`;

    static components = {
        'code-preview': CodePreview
    };

    initData() {
        return {
            sourceList: [{\\"filename\\":\\"index.ts\\",\\"code\\":\\"import {Component} from 'san';\\\\n\\\\nexport default class ComponentDemo extends Component {\\\\n    static template = &#96;\\\\n        &lt;div>\\\\n            &lt;button type=\\\\\\"button\\\\\\">First Button&lt;/button>\\\\n        &lt;/div>\\\\n    &#96;;\\\\n}\\",\\"type\\":\\"ts\\"}]
        };
    }
}
"
`;

exports[`transform component: custom preview template with more source 2`] = `
"import {Component} from 'san';
import CodePreview from '/home/work/demo/src/markdown/san-include-preview.md.Component2.vpms';

export default class PreviewBlock extends Component {

    static trimWhitespace = 'all';

    static template = \`
        <section class=\\"code-box\\" id=\\"2\\">
            <section class=\\"code-box-demo\\">
                <h4>code preview</h4>
                <code-preview></code-preview>
            </section>
            <section class=\\"code-preview-wrapper\\">
                <section
                    class=\\"code-preview\\"
                    s-for=\\"source in sourceList\\"
                >
                    <h3>{{source.filename}}</h3>
                    <code s-html=\\"source.code\\"></code>
                </section>
            </section>
        </section>
    \`;

    static components = {
        'code-preview': CodePreview
    };

    initData() {
        return {
            sourceList: [{\\"filename\\":\\"index.ts\\",\\"code\\":\\"import {Component} from 'san';\\\\n\\\\nexport default class ComponentDemo extends Component {\\\\n    static template = &#96;\\\\n        &lt;div>\\\\n            &lt;button type=\\\\\\"button\\\\\\">Second Button&lt;/button>\\\\n        &lt;/div>\\\\n    &#96;;\\\\n}\\",\\"type\\":\\"ts\\"}]
        };
    }
}
"
`;

exports[`transform component: custom preview template with more source 3`] = `
"import {Component} from 'san';
import CodePreview from '/home/work/demo/src/markdown/san-include-preview.md.Component3.vpms';

export default class PreviewBlock extends Component {

    static trimWhitespace = 'all';

    static template = \`
        <section class=\\"code-box\\" id=\\"3\\">
            <section class=\\"code-box-demo\\">
                <h4>code preview</h4>
                <code-preview></code-preview>
            </section>
            <section class=\\"code-preview-wrapper\\">
                <section
                    class=\\"code-preview\\"
                    s-for=\\"source in sourceList\\"
                >
                    <h3>{{source.filename}}</h3>
                    <code s-html=\\"source.code\\"></code>
                </section>
            </section>
        </section>
    \`;

    static components = {
        'code-preview': CodePreview
    };

    initData() {
        return {
            sourceList: [{\\"filename\\":\\"index.ts\\",\\"code\\":\\"import {Component} from 'san';\\\\nimport styles from '../index.module.less';\\\\n\\\\nexport default class ComponentDemo extends Component {\\\\n    static template = &#96;\\\\n        &lt;div class=\\\\\\"\${styles.content}\\\\\\">\\\\n            &lt;button type=\\\\\\"button\\\\\\">Third Button&lt;/button>\\\\n        &lt;/div>\\\\n    &#96;;\\\\n}\\",\\"type\\":\\"ts\\"},{\\"filename\\":\\"../index.module.less\\",\\"code\\":\\"\\",\\"type\\":\\"css\\"}]
        };
    }
}
"
`;

exports[`transform html 1`] = `
"<h1 id=\\"title\\">title</h1>
<p>123
456</p>
<pre><code class=\\"language-san\\">import {Component} from 'san';

export default SanDemo extends Component {
    static template = &#96;&lt;div>&lt;/div>&#96;;
    static components = {};
    initData () {
        return {
            foo: 'bar'
        };
    }
}</code></pre>"
`;
